src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -170,62 +170,86 @@ private static Task<Solution> GetSolutionAsync(RoslynServices roslynService, Pin
             return solutionController.GetSolutionAsync(solutionInfo.SolutionChecksum, solutionInfo.FromPrimaryBranch, cancellationToken);
         }
 
-        protected async Task<T> RunServiceAsync<T>(Func<Task<T>> callAsync, CancellationToken cancellationToken)
+        protected async Task<T> RunServiceAsync<T>(Func<CancellationToken, Task<T>> callAsync, CancellationToken cancellationToken)
         {
             AssetStorage.UpdateLastActivityTime();
 
-            try
+            // merge given cancellation token with shutdown cancellation token. it looks like if cancellation and disconnection happens
+            // almost same time, we might not get cancellation message back from stream json rpc and get disconnected
+            // https://github.com/Microsoft/vs-streamjsonrpc/issues/64
+            using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ShutdownCancellationToken))
```

Comment:
Are you aware of the dangers of disposing `CancellationTokenSource` objects?
It's the right thing for Linked CTS's, most of the time. But you should be aware of the risks if any of your CT's outlast your CTS.

Reply:
❓ Can you be more specific? Failure to dispose of `CancellationTokenSource` was a prominent source of memory leaks in the past, so we've moved to ensuring they are always disposed.

Reply:
Yes, you definitely want to avoid the memory leaks. The caution is around CT's that survive beyond the lifetime of their CTS. Seemingly benign calls like CT.Register and CT.ThrowIfCanceled() start throwing ObjectDisposedException after their CTS is disposed. I filed [a bug](http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems?id=369196&_a=edit) a long time back on this and it was fixed, but in a 'quirks' mode way. So I was never quite clear on who/where/how it may still bite.
I see the [relevant code in CoreCLR](https://github.com/dotnet/coreclr/blob/8a163ca04ae8734f4a9ee51e470beda933c96333/src/mscorlib/src/System/Threading/CancellationTokenSource.cs#L449) doesn't have this throwing behavior at all, but I'm not sure what is in NetFX at this point.

So morale of the story, unless you find out NetFX won't bite you, is make sure you await all your async work that accepts your CT, and then don't dispose the CTS until all awaited work is done.

Reply:
thank you Andrew. I dont think that will happen here but I will take a note.